core/src/eval/cache/lazy.rs

```diff
@@ -405,18 +405,50 @@ impl Thunk {
     }
 
     /// Set the state to evaluated.
-    pub fn set_evaluated(&mut self) {
+    pub fn set_evaluated(&self) {
         self.data.borrow_mut().state = ThunkState::Evaluated;
     }
 
+    /// Lock a thunk. This sets the thunk's state to [ThunkState::Blackholed]. If it's evaluated
+    /// while it's locked, an error will be raised by evaluation.
+    ///
+    /// Locking is used to prevent infinite loops for some step-by-step variants of deep evaluation
+    /// such as `%deep_seq%`, `%force%` or [crate::program::Program::eval_record_spine], where the
+    /// normal thunk update workflow isn't sufficient to detect infinite recursion. Otherwise, you
+    /// shouldn't use locking directly but rater [ThunkUpdateFrame] and [Self::mk_update_frame].
+    ///
+    /// # Preconditions
+    ///
+    /// **A thunk must be in the state [ThunkState::Evaluated]**. While this isn't a technical
+    /// limitation, we need to restore the previous state of the thunk upon unlocking. Since we
+    /// ever only lock evaluated thunks currently, it's simpler to always assume this is the case
+    /// and to restore the state to [ThunkState::Evaluated].
+    pub fn lock(&self) {
+        let mut data_ref = self.data.borrow_mut();
+        debug_assert!(data_ref.state == ThunkState::Evaluated);
+        data_ref.state = ThunkState::Blackholed;
+    }
+
+    /// Unlock a thunk previously locked (see [Self::lock]). If the thunk is locked or blackholed,
+    /// this method sets the state to [ThunkState::Evaluated]. Otherwise, it's a no-op.
+    pub fn unlock(&self) {
+        let mut data_ref = self.data.borrow_mut();
+
+        if let ThunkState::Blackholed = data_ref.state {
+            data_ref.state = ThunkState::Evaluated;
+        }
+    }
+
     /// Generate an update frame from this thunk and set the state to `Blackholed`. Return an
     /// error if the thunk was already black-holed.
     pub fn mk_update_frame(&mut self) -> Result<ThunkUpdateFrame, BlackholedError> {
-        if self.data.borrow().state == ThunkState::Blackholed {
+        let mut data_ref = self.data.borrow_mut();
```

Comment:
The changes in this function are cosmetic (avoid two calls to `borrow()`/`borrow_mut()` in a row: just borrow mut from the beginning)